CurrencyConverterControl/CurrencyConverterControl.cs

```diff
@@ -72,23 +101,25 @@ private static bool IsValidValue(object value)
         }
 
         // Adding a Text Change call back
-        private static void InputValue_Textchanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        private static void InputValue_TextChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
         {
             try
             {
                 CurrencyConverterControl control = (CurrencyConverterControl)d;
-                //check if all have values before going to API calculate
-                if (String.IsNullOrEmpty(control.SourceCurrency) || String.IsNullOrEmpty(control.DestinationCurrency))
-                {
-                    return;
-                }
-                control.OutputValue = CurrencyDataProvider.Convert(control.SourceCurrency,
-                                                            control.DestinationCurrency, control.InputValue);             
-              
+                control.PerformCalculation(control);
             }
             catch (Exception ex) { Console.WriteLine($"Error: {ex.Message}"); }
         }
 
+        // Define CLR events for input and output value changes
+        public event EventHandler<ValueChangedEventArgs> InputValueChanged;
+
+        public void OnInputValueChanged(object sender, ValueChangedEventArgs e)
```

Comment:
why it's public? And where it is called from? It's internal implementation detail, might be private method. InputValue_TextChanged should call this method to fire event